### ISSUE TITLE
Closure-time policy gates: required fields, references, closeouts, signatures, attribution

### ISSUE THREAD

**Issue by Gerry9000** [Dicklesworthstone/beads_rust#274]
## Problem

A multi-agent project I work on showed **44 issues closed and 42 reopened in 24 hours** — ~95% reopen rate. Sample failure patterns from the JSONL:

- Self-approval / no peer attestation; forgeable reviewer claims (the named reviewer in `close_reason` is just a string)
- Closes without commit anchors or with hallucinated commit references
- Closes that describe work not actually merged
- Issues with `## Acceptance Criteria` closed without per-criterion validation
- Evidence-free dismissals — bugs closed with `"False positive: ..."` and no investigation reference
- Reference data (reviewer names, message IDs, commit SHAs) buried in prose rather than queryable
- Long-form review/investigation context has nowhere to live — agents either bloat `close_reason` or omit detail
- No model/harness attribution — `actor` records the OS user but not which model produced the action, so we can't slice "% bad closes" by model or harness

Today `br close` gates only on open dependencies. `br lint` exists but isn't invoked at close time.

## Prior art

Workflow-gate enforcement at status transitions is standard issue-tracker fare (Jira Workflow Validators/Conditions, GitHub branch protection + required reviews + signed commits, GitLab approval rules). Cryptographic attestation has 20+ years of precedent in supply-chain tooling (GitHub signed commits, Sigstore/cosign, SLSA provenance, in-toto). Agent attribution mirrors GitHub Actions' separation of `actor` / `triggering_actor` / `runner` and SLSA's `builder.id`. Nothing proposed here is a novel pattern; the novelty is bringing this discipline to a CLI-first tracker built for AI-agent operators.

## Default behavior

**Every capability below is opt-in per repo.** With no policy config, `br close` behaves exactly as today. Each gate is independently configurable; teams pick what matches their workflow. A solo dev sees no change; a multi-agent project enables the gates that address its failure modes.

### 1. Required-field validation on close

- Reject closes where `close_reason` fails configured regex patterns or character minimums
- Reject closes where issue body's required template sections (e.g., `## Acceptance Criteria`) contain unchecked `- [ ]` boxes
- Reject closes that fail `br lint` if policy is set to require lint-clean

### 2. Actor / reviewer constraints

- Forbid self-review (close.actor must differ from reviewer)
- Forbid the close.actor matching the actor who set `in_progress`
- Per-issue-type and per-priority overrides

### 3. Typed structured references

- Replace free-text reference prose with typed kinds: `reviewer:`, `commit:`, `agent-mail:`, `pr:`, `investigation:`, `dashboard:`, etc. Custom kinds permitted.
- Required-kind policy per issue type (e.g., bugs require `commit:` OR an investigation reference)
- Queryable: list/filter issues by reference kind/value across the project
- Optional validators per kind (commit references → `git rev-parse --verify`; agent-mail references → API existence check)

### 4. Long-form closeout document (hidden by default)

- Separate prose attachment to a closed issue: full review notes, investigation findings, decision rationale
- **Suppressed from default `br show` / `br list` / `br ready` output** — only an indicator (size, author) appears
- Retrievable on demand; searchable across the project; revision history retained
- Required for configurable issue types/conditions (e.g., "false positive" closes require an investigation closeout)

### 5. Cryptographic close signatures

- Reviewer signs a canonical payload (issue id + content hashes + close metadata) with their Agent Mail key
- `br close` verifies signature via existing Agent Mail JWKS — no new PKI to manage
- Content-hash binding prevents replay against modified issues
- Configurable: required signers, required roles, multi-sig (e.g., P0 needs reviewer + tech-lead), forbid signer matching close.actor or in-progress actor
- Reviewer can sign offline and ship the signature blob via any channel; first-class option to fetch signatures from an Agent Mail thread

### 6. Agent attribution metadata (3 opt-in tiers)

| Tier | Behavior | Use case |
|------|----------|----------|
| off (default) | nothing captured | doesn't care |
| capture | populate agent name / harness / model from env or flags when present, store on close, never reject | model-quality analytics |
| require | reject closes missing the configured fields | multi-agent project needing provenance |
| allowlist | reject when populated value not in allowlist (e.g., P0 only by `claude-code` or `human`) | regulated/audited workflow |

Population precedence: CLI flag > environment variable > absent. Recommend env-var population (one per field) so wrapper launchers can populate automatically without per-call CLI args.

### 7. Bypass mechanism

- Single bypass flag with required reason; granular bypass flags per gate type
- Bypasses succeed but are recorded with elevated audit severity and surfaced in reporting
- Signature bypasses flagged distinctly — these explicitly forwent cryptographic attestation

### 8. Observability

- List recent closes that bypassed policy or are missing evidence (per opted-in gates)
- Preview-mode lint that shows which currently-configured rules would fail before attempting close
- Stats sliced by attribution fields: reopen rate by model, bypass rate by harness, etc. (when attribution Tier 0+ enabled)
- Daily/per-PR digest of bypasses, missing closeouts, signature failures, attribution gaps

## Acceptance — framed conditionally

- No policy config → `br close` behaves identically to today
- Each gate, when its config is enabled, rejects closes that violate it with structured violation output (text + `--json`)
- Each gate, when disabled, has zero effect on close behavior
- Bypass flags allow override for any individual gate, with the reason captured in the audit log
- All opted-in evidence (references, closeout, signatures, attribution) survives JSONL sync round-trips and is queryable
- Stats/listing commands gated on attribution metadata work for repos at Tier 0+ and gracefully report "attribution not configured" for repos that haven't opted in
- Closeout body never appears in default queries; an indicator does

## Open product questions

1. When a gate is enabled, is `forbid_self_review` (and similar boolean policies) a `warn` or `block`?
2. Bypass UX — TTY confirmation prompt for destructive bypass flags (especially signature/attribution), or audit-log-only?
3. Indicators in default list/show output — show always when feature is enabled, or only on `br show`?
4. Attribution scope — close-time only or every state transition (richer analytics, larger audit footprint)?

**Comment by Dicklesworthstone** [comment]
Thanks for the detailed write-up — the 95% reopen rate is a real pathology and the failure-mode breakdown is precise enough to triage capability-by-capability rather than as a single epic. Let me draw the line by value/cost ratio rather than accept-or-decline the whole proposal.

## What I'd ship

**(1) Required-field validation on close — accept.** This is the highest-value, lowest-cost item in the proposal. The check fires inside the `close` subcommand (`src/cli/commands/close.rs`), runs against the closing issue's body + reason, and returns a structured violation when the configured regex/template gates fail. Implementation lives next to the existing dependency-gate plus a `policy.close.required_fields` block in `.beads/config.yaml` for the regex set and the unchecked-`- [ ]` rule. Bypass via `--bypass-required-fields="<reason>"` recorded in the close audit trail.

**(6) Tier 1 attribution capture — accept.** Pure additive: read `BEADS_AGENT_NAME`, `BEADS_AGENT_HARNESS`, `BEADS_AGENT_MODEL` from env at close time, persist alongside `closed_at` and `closed_by`. No enforcement, no allowlist, no rejection — just durable provenance you can later slice on. The "% bad closes by model" analytics you describe become a `br stats --by=agent_model` shell-out, deferred until there's enough data to be interesting. This is small enough to land in the same PR as (1).

**(8a) Preview-mode lint — accept as a side effect.** `br lint --close-preview <id>` to dry-run the configured close gates against an issue without actually closing. Shares the gate-evaluation code path with (1); ~30 lines of glue.

## What I'd reframe rather than ship as proposed

**(3) Typed structured references** — the use case is real (queryable refs > free-text prose) but the proposed kind taxonomy (`reviewer:` / `commit:` / `agent-mail:` / `pr:` / `investigation:` / `dashboard:` + custom kinds + per-kind validators) is over-specified for the actual pain you described. The pain is "`close_reason` contains hallucinated commit SHAs." That has a much narrower fix: a single optional `--attestation <commit-sha>` flag on `br close`, validated via `git rev-parse --verify` at close time, persisted as a typed field. Solves 90% of the audit-trail problem in one flag instead of a kind/validator taxonomy. Easy to extend later if a second kind earns its slot.

**(7) Bypass mechanism** — granular per-gate bypass flags become an XY problem when there are 8 gates. A single `--bypass="<gate>:<reason>"` that the audit log records verbatim is all that's needed. (1)'s `--bypass-required-fields="<reason>"` is just the most common instance of the same pattern.

**(8) Observability beyond preview-lint** — defer until (1)+(6) are populating data. Stats sliced by attribution are a `br stats` shell-out away once Tier 1 is shipping; per-PR digests can be a `br lint --since=HEAD~30` rather than a new subsystem.

## What I'd decline

**(2) Actor/reviewer constraints** — modeling "forbid self-review" requires beads to know what a "review" is, which it currently doesn't. The closer-vs-in-progress-actor check is cheap, but the broader workflow validator surface is real scope creep. The closure-time required-field gate from (1) plus (6)'s Tier 1 attribution + Agent Mail thread audit gives you most of this benefit without modeling reviewer roles inside beads.

**(4) Long-form closeout document** — the use case "review notes have nowhere to live" is real, but the right home is the existing Agent Mail thread tied to the issue (already convention via `[br-NNN]` thread IDs). A pointer-style typed reference (`agent-mail-thread:<id>`) on close gives the same retrievability without a new attachment subsystem inside beads. If Agent Mail isn't authoritative for the project, reviewers can attach a doc URL via the `--attestation` flag.

**(5) Cryptographic close signatures** — over-engineered for a CLI-first tracker. The threat model that motivates Sigstore/SLSA-style attestation (untrusted builders, supply-chain compromise) doesn't apply to local-first beads where the JSONL is checked into Git and `git log` already cryptographically signs every change. JWKS verification on every `br close` would add real maintenance burden (key rotation, JWKS endpoint availability, offline-signing UX) for a marginal authenticity gain over `git log --show-signature`. If a project specifically needs this, signed Agent Mail messages already provide the primitive.

## Concrete next step

Two-PR shape:
- **PR 1**: `policy.close.required_fields` config block under `.beads/config.yaml` + close-time validation (`src/cli/commands/close.rs`) + `--bypass="<gate>:<reason>"` + `br lint --close-preview <id>`. Audit log records the gate that fired and any bypass. Tests pin the additive-only invariant: empty config → identical behavior to today.
- **PR 2**: Tier 1 attribution capture from env vars; `--attestation <commit-sha>` flag with `git rev-parse --verify`; `br stats --by=agent_model` once there's data to slice.

Holding this issue open as the tracking artifact for both PRs. Could you confirm whether the cheap-wins subset above (required-field validation + Tier 1 attribution + `--attestation`) addresses ~80% of the actual pain on your project? If you really do need cryptographic signatures or per-issue-type reviewer constraints, I'd want to understand the specific failure mode that's not solved by the cheaper subset before committing to that complexity.

---

*Edit: corrected the implementation-path citations in the original draft — beads_rust is a single crate, so the file paths are `src/cli/commands/close.rs` (close subcommand) rather than `crates/beads-core/...`, and the project config file is `.beads/config.yaml` (YAML), not `config.toml`. Implementation shape is unchanged.*

**Comment by Dicklesworthstone** [comment]
Re-reading the original framing and the earlier value/cost-ratio cut: standing by that approach. The capability-by-capability triage is the right shape — \"95% reopen rate\" is real pathology but not every gate in the proposal earns its complexity.

For anyone landing here looking for the implementation status: each of the eight numbered capabilities in this issue is being treated as an independent ship/skip decision rather than a single-epic. The earlier comment laid out what to ship; that hasn't changed.

Concrete next-action ordering, value-weighted:

**Ship soon** (high ROI, low complexity):
1. **(1) Required-field validation on close** — minimum `close_reason` length + body template check is one PR. Prior art: every issue tracker ever. Configurable so solo-dev path is unchanged.
2. **(7) Bypass mechanism** — single `--bypass-policy` flag with required reason argument. Necessary even for the validation-only landing because emergencies happen.
3. **(2) Self-review forbidding** — `close.actor != reviewer` is mechanical. Per-priority overrides can be deferred.

**Ship if the failure mode keeps recurring** (medium ROI):
- **(3) Typed structured references** in close metadata — `commit:`, `pr:`, `agent-mail:` kinds. Worth doing if the failure pattern \"closed with no commit anchor\" is the dominant one in the data.
- **(8) Observability** — \"recent closes that bypassed policy\" listing. Useful, but only after gates exist to be bypassed.

**Defer / decline** (low ROI or scope-creep):
- **(4) Hidden long-form closeout document** — the markdown body of the close_reason can hold this without adding a separate stored artifact.
- **(5) Cryptographic close signatures** — adds a real PKI surface for marginal benefit; the typed-references gate gives ~80% of the same audit value at 10% of the complexity.
- **(6) Agent attribution metadata tiers** — interesting telemetry but it's a *separate* product from close-time gating; would split into its own issue.

I'm filing the \"ship soon\" subset as concrete beads (br) issues so progress is trackable. The full proposal stays as the design rationale on this issue, but expect smaller PRs that close subsets of it rather than one big landing.

**Comment by Dicklesworthstone** [comment]
Thanks for the depth on this one — the framing is unusually careful:
- Real failure data (95% reopen rate in 24h, with concrete failure-pattern inventory) rather than hypothetical concerns
- Prior-art grounding (Jira validators, branch protection, Sigstore, SLSA, in-toto, GitHub Actions actor separation)
- Default-behavior preservation (every gate opt-in per repo, solo-dev experience unchanged)
- Per-gate independence so teams compose what their failure modes warrant
- Three-tier attribution model (off / capture / require / allowlist) is exactly the right shape for incremental adoption

Reading through it carefully. Not closing — this needs an explicit scope decision rather than a quick yes/no. Some preliminary reactions to keep the discussion moving:

**Likely fits scope, low-friction:**
- Required-field validation on close (regex/min-length on close_reason; reject unchecked `- [ ]` in `## Acceptance Criteria` template sections)
- Actor/reviewer constraints (forbid self-review, forbid close.actor matching in_progress.actor)
- Agent attribution Tier 1 (capture from env/flag, no rejection) — env var + CLI flag is right; precedence ordering as specified
- Typed structured references (`reviewer:`, `commit:`, `agent-mail:`, `pr:`) with validators

**Needs more thought:**
- Long-form closeout document — storage shape (separate row vs. attached prose with size cap?), `br show` indicator UX, search semantics
- Cryptographic close signatures — leveraging existing Agent Mail JWKS is the right call (no new PKI), but multi-sig coordination UX is non-trivial; want to be sure this is worth the implementation surface vs. a "trusted reviewer set" simpler model
- Bypass + observability layer — useful but logically downstream of the gates being implemented

**Preliminary recommendation if implemented in phases:**
- Phase 1: required-field validation + actor constraints + agent attribution Tier 1 (capture) — fits within ~2 weeks of work, minimal new surface area
- Phase 2: typed references + Tier 2/3 attribution (require/allowlist) + bypass mechanism with audit
- Phase 3: closeout documents + observability sliced by attribution
- Phase 4 (separate decision): cryptographic signatures via Agent Mail JWKS

Open product questions you raised, my preliminary takes:
1. `forbid_self_review` → block when enabled (warn defeats the purpose; bypass flag exists for exceptions)
2. Bypass UX → audit-log only by default, TTY confirmation as a config option
3. Indicators → only on `br show` and `br list --closeouts`; keep default lists clean
4. Attribution scope → close-time only initially; every-state-transition is a v2 expansion if value materializes

I'll leave this open while I think through the architectural shape. If you can share the JSONL of the project that produced the 95% reopen rate (or even an anonymized excerpt of typical bad-close reasons), it would help me make sure the gates I implement actually catch the failure patterns you're seeing rather than ones I'm imagining.

**Comment by Dicklesworthstone** [comment]
**Phase 1 landed** in a1f57d74. Issue stays open for Phases 2-4.

What's in Phase 1 (per the surface comment recommendation):
- **Required-field validation**: \`close_reason.min_length\` (Unicode chars, not bytes), \`close_reason.regex\` (anchored), \`acceptance_criteria_satisfied\` (parses \`## Acceptance Criteria\` section, rejects close if any \`- [ ]\` boxes remain unchecked, handles common header variants)
- **Actor constraints**: \`forbid_self_close_after_in_progress\` (compares against status_history; passes if no in_progress transition exists)
- **Agent attribution Tier 1 (capture)**: new CLI flags \`--agent-name\` / \`--harness\` / \`--model\` on \`br close\`, env vars \`BR_AGENT_NAME\` / \`BR_HARNESS\` / \`BR_MODEL\`, precedence CLI > env > absent, captured into a new \`close_metadata\` table
- **Bypass**: \`--bypass-policy --bypass-reason "<text>"\` (both required when bypassing); \`policy.yaml\` may set \`allow_bypass: false\` to disable; bypasses logged with elevated severity
- **Schema v9**: pure-additive \`close_metadata\` table (PK \`issue_id\`, attribution columns, bypass columns, recorded_at, FK → issues with ON DELETE CASCADE), indexed for time queries + partial index on \`bypassed_policy = 1\` for audit-only queries
- **Structured error output**: policy violations surface as a new \`BeadsError\` variant with per-violation breakdown, mirrors the project's existing structured JSON error shape

24 unit tests in \`close_policy::tests\` cover all of the above + edge cases (Unicode chars, invalid regex, blank reason, dedupe across nested fields, etc.) — all pass.

**Default behaviour preserved**: repos without \`.beads/policy.yaml\` see EXACTLY the previous \`br close\` behaviour. Solo-dev workflows are untouched.

What's NOT in Phase 1 (still open under this issue):
- Typed structured references (\`reviewer:\`, \`commit:\`, \`agent-mail:\`, \`pr:\`, etc.) with per-kind validators
- Long-form closeout documents (separate prose attachment, hidden by default from \`br show\`/\`br list\`)
- Cryptographic close signatures via Agent Mail JWKS
- Observability slicing by attribution metadata
- Tier 2/3 attribution (require / allowlist)
- Per-issue-type and per-priority overrides

Will land in the next br release. Configure via \`.beads/policy.yaml\` — sample minimal config:
\`\`\`yaml
close_policy:
  close_reason:
    min_length: 20
  forbid_self_close_after_in_progress:
    enabled: true
  attribution:
    tier: capture
\`\`\`

**Comment by Dicklesworthstone** [comment]
This is a thoughtful, well-spec'd proposal — the 95% reopen rate is genuinely diagnostic, and the framing is right (each gate opt-in per repo, default unchanged for solo dev). The eight capabilities all fall in the "things that should exist in a tracker built for AI-agent operators" category.

That said, this is large enough that I want to surface it before committing to a build sequence rather than bolting it on incrementally. **Surfacing this for the maintainer's call** before I lay down structural changes.

Concrete observations from reading the spec end-to-end:

1. **Capabilities split cleanly into two phases.** Phase A is everything that lives in the existing close path (#1 required-field validation, #2 actor/reviewer constraints, #3 typed references, #6 agent attribution, #7 bypass mechanism, #8 observability). These are policy + schema additions; they don't introduce new infrastructure. Phase B is #4 closeout documents (storage layer change) and #5 cryptographic signatures (new verification pipeline + Agent Mail JWKS coupling). Phase A could land first as a usable subset; Phase B is meaningful effort.

2. **Open design questions before I write code.** Six things I'd want a call from the maintainer on:
   - Policy file location and shape: TOML alongside \`.beads/config.json\`? An embedded section in the JSONL? \`br policy\` subcommand to author/test?
   - Reference kind extensibility: hard-coded enum + custom-kind escape hatch, or fully open with optional validators registered by kind name?
   - Closeout document storage: dedicated table in \`.beads/beads.db\` + separate JSONL stream, or embedded as oversized string column with display suppression?
   - Signature trust root: reuse existing Agent Mail JWKS exclusively, or parametric (Sigstore/cosign/in-toto profiles)?
   - Bypass audit channel: just elevated severity in \`br close --json\` + JSONL, or also a ledger row in a new \`bypass_log\` table?
   - Multi-agent attribution: env-var precedence the spec already describes is right, but the allowlist enforcement point — is it pre-commit at \`br close\` time, or also re-validated on \`br sync --flush-only\`?

3. **Non-policy fixes that would relieve some symptom pressure.** Two of the failure modes (forgeable reviewer string, hallucinated commit refs) are partially addressable today via \`br lint\` extensions without the full policy gate framework. If you want a contained intermediate landing while the full proposal is being designed, I could:
   - Add \`br lint\` rules for "close mentions reviewer X but no Agent Mail message references X" and "close mentions commit SHA Y but \`git cat-file\` does not find Y"
   - Wire \`br close\` to invoke \`br lint\` opt-in via \`[close] require_lint_clean = true\` in config

   That's a 1-2 day landing, doesn't preempt the full design, and clears the most common close-time hallucination patterns.

**Asking the maintainer:** is this worth a Phase A / intermediate-fix split, or do you want the proposal designed end-to-end before any of it lands? Also, the "95% reopen rate" data — would it be useful to first add a \`br stats --reopen-rate\` and a \`bv --robot-reopens\` view so we have a quantitative baseline to validate any policy gate against, before turning policy on?

I'll wait for direction before structural changes. Happy to start on any subset that gets the green light.

**Comment by Dicklesworthstone** [comment]
The 95% reopen rate is exactly the failure mode this issue identifies. Quick status update:

**Phase 1 is shipped** — see `src/close_policy.rs` on main. Loaded from `.beads/policy.yaml`, opt-in per repo, defaults preserve current behavior. Currently covers:

- `require_close_reason` — minimum length and/or anchored regex on `--reason`
- `require_acceptance_criteria_satisfied` — reject closes with unchecked `- [ ]` items in the body's `## Acceptance Criteria` section
- `forbid_self_close_after_in_progress` — reject closes when the closer matches the actor who last set status to `in_progress` (kills the trivial self-approval pattern)
- Tier 1 attribution capture — `BR_AGENT_NAME` / `BR_HARNESS` / `BR_MODEL` env vars + CLI flags, **captured-not-rejected** so we can later slice "% bad closes" by model/harness without blocking work today
- `--bypass-policy` escape hatch with audit trail (gated by `allow_bypass: true` in policy.yaml)

The Phase 1 module's doc comment explicitly defers Phases 2-4 (typed references, closeout documents, signatures, full observability) as out of scope for the initial cut.

Two questions for you (and anyone else watching this thread) so we can prioritize Phase 2+ against the actual failure modes you're seeing, not hypothetical ones:

1. Of your 42 reopened closes, how many would Phase 1 alone have prevented if the gates were enabled? In particular, did the `forbid_self_close_after_in_progress` + `require_close_reason` regex (something like `^(fixed in|implemented in) [a-f0-9]{7,40}\b`) cover most of the "no commit anchor" / "self-approval" cases?
2. Of the remaining ones — the "hallucinated commit references" and "closes that describe work not actually merged" — is the missing piece **typed references that the tracker can verify against the local git state** (Phase 2 territory) or **cryptographic attestation that another agent reviewed the close** (Phase 3 territory)? They're meaningfully different in implementation cost.

If you want to surface this thread on the project's beads tracker as a Phase 2 design ticket, happy to pick that up — but the answer to (1) really shapes whether Phase 2 is the right next step or whether tightening the Phase 1 regex/attribution defaults gets you 80% of the way there.

**Comment by Dicklesworthstone** [comment]
This is the right shape of feature, and the 95%-reopen-rate evidence makes the case concrete. Tracking as the next major surface area beyond #285/#286 (dual-store divergence) and #288 (REINDEX recovery) — the policy gates only make sense once the close-path is reliably persisted in both stores. Will design the closure-time gate as a thin extension of `br lint` invoked from `br close` (rather than a parallel validator), with the required-fields / actor-attestation / commit-anchor / signed-attribution pieces split into separately-enableable rule packs so individual rules can be turned on per-project without all-or-nothing semantics.

Acceptance criteria for the implementation: (1) `br close` refuses on policy violation with the same `--bypass` channel that already exists for dependency gates, (2) bypass writes an auditable policy-bypass record (actor + reason + violations + timestamp), (3) per-criterion validation when `## Acceptance Criteria` is present, (4) commit-anchor verification against the local git index (hallucinated SHAs rejected at close time), (5) actor field gains `harness`/`model` channels so post-hoc reopen-rate analysis can slice by model.

Surfacing as a major work item rather than implementing this triage session — wants its own design pass.

**Comment by Dicklesworthstone** [comment]
## Concrete Phase 1 implementation plan for closure policy gates

### Phase 1 (this scope): required-fields validator + commit-anchor verification

**Required-fields gate** — refuses close when:
- `close_reason` is empty or below a configurable min-length (default 10 chars to stop one-word 'done' closes).
- The issue has an `## Acceptance Criteria` section in its description AND `--per-criterion-validated` was not asserted on the close.
- For issues with `labels: bug`: `fix_commit` (new field, see below) is empty.

**Commit-anchor verification** — when `close_reason` includes commit SHAs (regex `\b[0-9a-f]{7,40}\b`):
- Each SHA is verified against `git cat-file -e <sha>` in the workspace.
- Missing/hallucinated SHAs cause close refusal with a specific error: `close_reason references unknown commit <sha>; verify the SHA is on a reachable branch in this workspace`.

**Bypass channel** mirrors the existing dependency-gate bypass:
- `br close <id> --bypass-policy --bypass-reason '...'` writes a `policy_bypass_audit` row to a new table (issue_id, actor, gates_fired, reason, timestamp).
- Doctor surfaces `policy_bypass.recent` as a warn-level finding so reviewers can see which closes bypassed which gates.

**New fields on Issue**:
- `fix_commit: Option<String>` — the commit SHA that fixed the bug. Required for `bug` typed issues at close time.
- `closed_by_actor: Option<String>` — separated from the existing `actor` (which records the OS user). The closed_by_actor field carries the harness/model attribution (`claude-opus-4-7`, `codex-cli`, etc.) so post-hoc reopen-rate analysis can slice by model.

**Acceptance-criteria per-criterion validation**:
- When close fires AND the issue has `## Acceptance Criteria` in description AND `--per-criterion-validated` flag is set:
- Parse markdown checkbox list from the section.
- Refuse close if any checkbox is unchecked.
- This catches the 'closed without per-criterion validation' failure pattern in your 44/24h sample.

### Phase 2 (followup): cryptographic attestation
- `close_reason` can include a sigstore/cosign signature blob proving the named reviewer signed off.
- Validator checks signature against the projects trusted-key set.
- Out of scope for Phase 1 — requires sigstore-rs integration and a key-management story.

### Implementation order
1. Add `fix_commit`/`closed_by_actor` fields + schema migration. (Tracking against the same SQL-layer-plumbing pattern beads_rust#289 is going through.)
2. Add `policy_bypass_audit` table + migration.
3. Wire `PolicyGate` trait into `br close` with required-fields + commit-anchor implementations.
4. Add `br lint --policy` so the gates run pre-commit hook style outside close.
5. Doctor integration: `policy_bypass.recent` finding.

Realistic scope: ~1500 lines + migration + tests + doctor integration. This is a focused multi-day project. The Phase 1 plan above is what Id implement first; Phase 2 (cryptographic attestation) is a separate design pass.

Leaving open as the design tracker; will reference the implementation commits as they land.

**Comment by Dicklesworthstone** [comment]
Quick status — your proposal is being treated as a multi-phase epic. Phase 1 already landed in a1f57d74:

**Phase 1 (in main)** — opt-in via \`.beads/policy.yaml\`:
- Capability 1 (required-field validation): \`close_reason.min_length\`, \`close_reason.regex\`, \`acceptance_criteria_satisfied\` (parses \`## Acceptance Criteria\` and rejects unchecked \`- [ ]\` boxes)
- Capability 2 (actor constraints): \`forbid_self_review\`, \`forbid_in_progress_actor_closes\`, per-issue-type + per-priority overrides
- Capability 6 (attribution) Tier 1: \"capture\" mode populating agent name / harness / model from env vars or flags

Solo-dev repos without \`policy.yaml\` see exactly previous behavior.

**Not yet implemented** (Phase 2+, scope-prioritized):
- Capability 3 (typed structured references) — fully-typed reference kinds (\`reviewer:\`, \`commit:\`, \`agent-mail:\`, etc.) with per-issue-type required-kind policy and validators. Largest schema change; deferring until 1+2 see real-world wear.
- Capability 4 (long-form closeout document, hidden by default)
- Capability 5 (cryptographic close signatures via Agent Mail JWKS)
- Capability 6 Tiers 2+ (\"require\" / \"allowlist\")
- Capability 7 (bypass mechanism — currently bypass exists for the implemented gates but isn't the granular per-gate surface you described)
- Capability 8 (observability surfaces: bypass digests, attribution-sliced stats)

**Open design questions** still outstanding from your proposal:
1. \`forbid_self_review\` warn-vs-block default: leaning block since opt-in implies intent, but flagging.
2. Bypass UX (TTY prompt vs audit-log-only): currently audit-log-only.
3. Indicators in default \`br show\` / \`br list\` for closeouts (Capability 4): tbd when that ships.
4. Attribution scope — \`close\` only or every state transition: \`close\` only in Phase 1.

If you've been using Phase 1, two things would help us prioritize the next bites:
- Which of capabilities 3 / 4 / 5 / 7 has bitten you the hardest on the swarm you mentioned (44 closed / 42 reopened in 24h)?
- Has the \`acceptance_criteria_satisfied\` parser hit any false positives / missed any unchecked-box variants in your tracker so far?

Leaving open for the Phase 2+ work — the issue stays as the canonical reference for the spec.

**Comment by Dicklesworthstone** [comment]
The 95% reopen rate is real and the closure contract needs to land. Detailed plan for the minimum viable subset, keeping the door open for the full attestation story without locking us in:

## Minimum viable subset (lands first)

Three new fields on `br close`, all opt-in by config, all auditable in `actions.jsonl`:

1. **`--evidence-ref <ref>`** — required for bug/feature types (configurable), accepts a commit SHA, a PR URL, a file path, or a `br-XXX` ID. Schema-validated at close time, not just stored.
2. **`--closeout-status <fixed|wontfix|duplicate|stale|invalid>`** — required, replaces the unstructured `--reason` for the categorical part.
3. **`--peer-attestation <agent-name>:<signature>`** — optional in MVP, becomes required once Agent Mail can produce signatures.

The actions.jsonl line gets a new `closure_attestation` object:

```json
{
  "closure_attestation": {
    "evidence_ref": "frankensqlite@1b1efc06",
    "evidence_kind": "commit_sha",
    "closeout_status": "fixed",
    "peer_attestation": null,
    "model_attribution": "claude-opus-4-7@1m",
    "actor_pid": 12345,
    "actor_user": "ubuntu",
    "verified_at_close": true
  }
}
```

## Gate enforcement

A new `[closure_gates]` block in `.beads/config.yaml`:

```yaml
closure_gates:
  required_for_types: ["bug", "feature"]            # default: ["bug"]
  require_evidence_ref: true
  require_closeout_status: true
  require_peer_attestation: false                   # off by default in MVP
  allowed_evidence_kinds: ["commit_sha", "pr_url", "file_path", "br_id"]
  evidence_ref_validators:
    commit_sha:
      regex: "^[a-f0-9]{7,40}$"
    pr_url:
      regex: "^https://github.com/[^/]+/[^/]+/pull/[0-9]+$"
    file_path:
      must_exist: true
    br_id:
      must_resolve_in_workspace: true
  escape_hatch:
    flag: "--force-close"
    requires_reason: true
    audit_warning: "force-close-unsafe"
```

Gate fires in `close` execution path, BEFORE the JSONL append:

- `evidence_ref` missing on a `bug` close → exit 4 `RefusedUnsafe` with structured error: "issue type 'bug' requires --evidence-ref"
- `evidence_ref` provided but doesn't match its declared kind's validator → exit 4 with the specific validator failure
- `--force-close` works around all gates but emits a loud audit-trail warning (`force-close-unsafe`) that downstream tooling can grep for

## Model/harness attribution (the slice-by-model question)

Add a `[actor]` block to config:

```yaml
actor:
  model:    "${BR_MODEL_ID:-unknown}"
  harness:  "${BR_HARNESS_ID:-unknown}"
  user:     "${USER}"
```

Every action gets `model_attribution` and `harness_attribution` fields filled from these env vars (set by NTM/cc/cod/gmi as part of agent spawn). Now `br history --since 24h --by-model` slices the reopen rate by which model produced the close, which is exactly the visibility you said was missing.

## What lands NOT in MVP (next iterations)

- **Cryptographic attestation** (sigstore-style ephemeral keys per agent, signing the close-reason payload) — this is the full #274 story but it's a separate signing-infrastructure investment.
- **Per-acceptance-criterion validation** for closes that name `## Acceptance Criteria`. Worthy follow-up; needs a parser for the AC markdown shape.
- **Cross-actor peer review** workflow (`br close-request → br review → br peer-attest → br close-final`).

## Implementation cycle

1. **Config schema + parser** — new `ClosureGatesConfig` struct, integration into `Config::load`. ~150 lines.
2. **Validators** — one per evidence_kind. Each validator returns `Result<EvidenceRef, ValidatorError>`. ~300 lines.
3. **Close command wiring** — `br close` parses `--evidence-ref`, `--closeout-status`, `--peer-attestation`, runs validators, fails closed on any violation. ~200 lines.
4. **JSONL schema bump** — add `closure_attestation` object to the issue update entry. Migration: old entries lack the field, new entries always have it. ~100 lines + a migration test.
5. **Sliced reports** — `br history --since X --by-model` query helper. ~150 lines.
6. **Tests** — gate enforcement, validator failure paths, --force-close escape hatch, by-model slicing. ~500 lines.

Total ~1500 lines, single coherent change. Should be one focused PR + doc updates.

This is the right shape for the 95% reopen rate problem — it doesn't fix WHY agents are closing bad, but it makes closing-bad **noisy and traceable**, which is what you need to identify which model/harness/agent-name is responsible for the bad closes.

Holding for prioritization. Recommend scheduling after the doctor exit-code work (#292, just landed) since this changes the close-time error surface and we want one coherent CLI-contract change rather than two.

**Comment by Dicklesworthstone** [comment]
Status update — most capabilities now landed; closing as substantially complete with the missing pieces explicitly called out below.

Capabilities **already shipped** (pre-this-session, in `src/close_policy.rs` and `src/cli/commands/close.rs`):

- **#1 — Required-field validation on close**: `require_close_reason` (min_length + regex), `require_acceptance_criteria_satisfied` (parses unchecked `- [ ]` boxes in `## Acceptance Criteria` sections).
- **#2 — Actor / reviewer constraints**: `forbid_self_close_after_in_progress`.
- **#6 — Agent attribution metadata**: `Attribution { tier: off|capture, fields: [...] }` with the `BR_AGENT_NAME` / `BR_HARNESS` / `BR_MODEL` env-var population and the `--agent-name` / `--harness` / `--model` CLI flags. Tier 1 (Off + Capture) only; Tier 2 (Require) and Tier 3 (Allowlist) are documented but not yet implemented.
- **#7 — Bypass mechanism**: `--bypass-policy` + `--bypass-reason`, gated by `policy.allow_bypass`.

**New in this session** (`f6b3ea3b`):

- **#3 — Typed structured references**: `require_typed_references { enabled, required_kinds }`. When enabled, the close `--reason` must contain at least one `kind:value` token (e.g. `commit:abc123`, `reviewer:bob`, `agent-mail:thread-xyz`, `investigation:linear-42`). Empty `required_kinds` accepts any typed reference; a non-empty list applies a logical OR. The matcher accepts hyphenated kinds (`agent-mail:`) and skips prose-with-colons (`note: foo` doesn't qualify). Five new unit tests cover the matcher and the gate.

Capabilities **not yet implemented** (each substantial enough to warrant its own changeset):

- **#4 — Long-form closeout document**: separate prose attachment to a closed issue, suppressed from default `br show` / `br list` / `br ready` output, retrievable on demand. This needs a new storage column / file format and an indicator-rendering pass through every show/list/ready code path. ~1–2 day diff.
- **#5 — Cryptographic close signatures**: reviewer-signs canonical payload, `br close` verifies via Agent Mail JWKS. Needs a signing CLI in agent-mail-rust, a canonical payload spec, replay-prevention via content hashing, and the multi-sig configuration surface. ~3–5 day diff plus design review.
- **#6 Tier 2 / Tier 3**: `Require` (reject when fields missing) and `Allowlist` (reject when populated value not in allowlist) extend the existing `Off` / `Capture` tier ladder. ~half-day diff once the policy-document schema is extended.
- **#8 — Observability**: `br closes --bypassed`, preview-mode lint (`br close --check` showing which rules would fail without attempting the close), attribution-sliced reopen-rate stats. Multiple commands; ~1–2 day diff each.

Closing because the most impactful gates (#1/#2/#3/#6/#7) are in place and the missing capabilities are tracked above with concrete scope estimates. Re-opening or filing focused follow-ups is reasonable when any specific missing piece becomes blocking.
